<a href="https://colab.research.google.com/github/Napassornmp/Project_CUSTOMER-CHURN-K-Means/blob/percentage/Project_CUSTOMER_CHURN_K_Means_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Identifying High-Risk customers likly to churn Using K - Means Clustering
  The purpose is to identify customers in  High-Risk churn cluster and estimate the size of the potential churn group.

  Implement strategies to retain at-risk  customers.

This study aims to explore which attributes in each class have similar value patterns and which attributes have the greatest influence on the class.

In [1]:
import pandas as pd

In [2]:
#loading data
df = pd.read_csv('/content/drive/MyDrive/DataforProject/WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [3]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [4]:
df = df.drop(['customerID'], axis=1)


df = ตัวแปรอิสระ

In [5]:
df

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0                 Yes           Electronic check           29.85        29.85   
1                  No               Mailed check           56.95       1889.5   
2                 Yes               Mailed check           53.85       108.15   
3                  No  Bank transfer (automatic)           42.30      1840.75   
4                 Yes           Electronic check           70.70       151.65   
...               ...                        ...             ...          ...   
7038              Yes               Mailed check           84.80       1990.5   
7039              Yes    Credit card (automatic)          103.20       7362.9   
7040              Yes           Electronic check           29.60       346.45   
7041              Yes               Mailed check           74.40        306.6   
7042              Yes  Bank transfer (automatic)          105.65       6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 20 columns]

In [6]:
# กรองเฉพาะ Churn = 'Yes'
df_yes = df[df['Churn'] == 'Yes']

# นับจำนวนแถว
count_yes = df_yes.shape[0]
print(f"จำนวนตัวอย่างที่ Churn = Yes: {count_yes}")


จำนวนตัวอย่างที่ Churn = Yes: 1869


In [7]:
# กรองเฉพาะ Churn = 'Yes'
df_No = df[df['Churn'] == 'No']

# นับจำนวนแถว
count_No = df_No.shape[0]
print(f"จำนวนตัวอย่างที่ Churn = No: {count_No}")

จำนวนตัวอย่างที่ Churn = No: 5174


# Data is imbalance >> Sampling by Random Oversampling

In [8]:
df_yes  = df[df['Churn']=='Yes']
df_no   = df[df['Churn']=='No']


y = ตัวแปรตาม Churn

X_res =  sampling independent

y_res = sampling Dependent

In [9]:
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

y = df['Churn']
# สร้าง undersampler ให้ลด “No” ให้เท่ากับ “Yes”
rus = RandomUnderSampler(sampling_strategy='majority', random_state=42)

X_res, y_res = rus.fit_resample(df, y)

print("Before:", Counter(y))
print("After:", Counter(y_res))

Before: Counter({'No': 5174, 'Yes': 1869})
After: Counter({'No': 1869, 'Yes': 1869})


In [10]:

print("Before:", Counter(y))
print("After:", Counter(y_res))


Before: Counter({'No': 5174, 'Yes': 1869})
After: Counter({'No': 1869, 'Yes': 1869})


In [11]:
X_res.nunique() #Checking Unique values

gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1297
TotalCharges        3500
Churn                  2
dtype: int64

In [12]:
X_res.isnull().sum() #Check null values

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [13]:
X_res.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [14]:
X_res.shape
y_res.shape

(3738,)

# เจอคอลัมน์ที่ “น่าจะเป็นตัวเลข แต่เป็น object” เช่น 'TotalCharges'
ให้ใช้ pd.to_numeric(..., errors='coerce') เฉพาะคอลัมน์นั้น เพื่อตรวจสอบเนื่องจาก
Object มักซ่อนค่าแปลกๆไว้

In [15]:
df_totalCharges = pd.to_numeric(X_res.TotalCharges, errors='coerce')
X_res.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Show that 11 rows

In [16]:
invalid_rows = X_res[X_res['TotalCharges'].isna()]

# แสดงผล
invalid_rows


Empty DataFrame
Columns: [gender, SeniorCitizen, Partner, Dependents, tenure, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, PaymentMethod, MonthlyCharges, TotalCharges, Churn]
Index: []

Replace column 'TotalCharges' with NaN by 0 .

In [17]:
X_res['TotalCharges'] = X_res['TotalCharges'].fillna(0)

# ตรวจสอบผล
df['TotalCharges'].isna().sum()  # จะได้ 0

np.int64(0)

## Category and Numeric Columns


In [18]:
X_res.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

df_features = sampling independent rows without class

In [19]:
# แยก target และตัด ID ออก
df_features = X_res.drop(['Churn'], axis=1)

# เลือกเฉพาะ categorical columns
CategoriesCol = [ 'gender', 'Partner', 'Dependents', 'PhoneService',
       'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'Contract', 'PaperlessBilling', 'PaymentMethod']

# ทำ dummy
df_dummy = pd.get_dummies(df_features, columns=CategoriesCol, drop_first=False)



In [20]:
print(len(df_dummy), len(X_res))


3738 3738


Manipulate Categorical Column usinh Dummy.

In [21]:
df_dummy = pd.get_dummies(df_features, columns= CategoriesCol, drop_first=False)


In [22]:
# เพิ่ม target กลับ
df_dummy['Churn'] = X_res['Churn']


In [23]:
X_res.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [24]:
df_dummy.columns


Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges',
       'gender_Female', 'gender_Male', 'Partner_No', 'Partner_Yes',
       'Dependents_No', 'Dependents_Yes', 'PhoneService_No',
       'PhoneService_Yes', 'MultipleLines_No',
       'MultipleLines_No phone service', 'MultipleLines_Yes',
       'InternetService_DSL', 'InternetService_Fiber optic',
       'InternetService_No', 'OnlineSecurity_No',
       'OnlineSecurity_No internet service', 'OnlineSecurity_Yes',
       'OnlineBackup_No', 'OnlineBackup_No internet service',
       'OnlineBackup_Yes', 'DeviceProtection_No',
       'DeviceProtection_No internet service', 'DeviceProtection_Yes',
       'TechSupport_No', 'TechSupport_No internet service', 'TechSupport_Yes',
       'StreamingTV_No', 'StreamingTV_No internet service', 'StreamingTV_Yes',
       'StreamingMovies_No', 'StreamingMovies_No internet service',
       'StreamingMovies_Yes', 'Contract_Month-to-month', 'Contract_One year',
       'Contract_Two year', '

In [25]:
df_dummy

SeniorCitizen  tenure  MonthlyCharges TotalCharges  gender_Female  \
2134              0       1           20.15        20.15           True   
2693              0      72           19.30       1414.8          False   
1679              1       7           45.75        344.2          False   
5998              0      48           20.05         1036          False   
6798              0      13           70.15        886.7           True   
...             ...     ...             ...          ...            ...   
7021              0      12           59.80        727.8          False   
7026              0       9           44.20       403.35           True   
7032              1       1           75.75        75.75          False   
7034              0      67          102.95      6886.25           True   
7041              1       4           74.40        306.6          False   

      gender_Male  Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  \
2134        False        True        False           True           False   
2693         True        True        False           True           False   
1679         True        True        False          False            True   
5998         True       False         True           True           False   
6798        False        True        False           True           False   
...           ...         ...          ...            ...             ...   
7021         True        True        False           True           False   
7026        False        True        False           True           False   
7032         True        True        False           True           False   
7034        False        True        False           True           False   
7041         True       False         True           True           False   

      ...  Contract_Month-to-month  Contract_One year  Contract_Two year  \
2134  ...                     True              False              False   
2693  ...                    False              False               True   
1679  ...                     True              False              False   
5998  ...                    False              False               True   
6798  ...                     True              False              False   
...   ...                      ...                ...                ...   
7021  ...                    False               True              False   
7026  ...                     True              False              False   
7032  ...                     True              False              False   
7034  ...                     True              False              False   
7041  ...                     True              False              False   

      PaperlessBilling_No  PaperlessBilling_Yes  \
2134                 True                 False   
2693                 True                 False   
1679                False                  True   
5998                False                  True   
6798                False                  True   
...                   ...                   ...   
7021                False                  True   
7026                False                  True   
7032                False                  True   
7034                False                  True   
7041                False                  True   

      PaymentMethod_Bank transfer (automatic)  \
2134                                    False   
2693                                     True   
1679                                    False   
5998                                    False   
6798                                    False   
...                                       ...   
7021                                    False   
7026                                     True   
7032                                    False   
7034                                    False   
7041                                    False   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electr

In [26]:
columns_list = df_dummy.columns.tolist()
print(columns_list)
print(len(columns_list))

['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges', 'gender_Female', 'gender_Male', 'Partner_No', 'Partner_Yes', 'Dependents_No', 'Dependents_Yes', 'PhoneService_No', 'PhoneService_Yes', 'MultipleLines_No', 'MultipleLines_No phone service', 'MultipleLines_Yes', 'InternetService_DSL', 'InternetService_Fiber optic', 'InternetService_No', 'OnlineSecurity_No', 'OnlineSecurity_No internet service', 'OnlineSecurity_Yes', 'OnlineBackup_No', 'OnlineBackup_No internet service', 'OnlineBackup_Yes', 'DeviceProtection_No', 'DeviceProtection_No internet service', 'DeviceProtection_Yes', 'TechSupport_No', 'TechSupport_No internet service', 'TechSupport_Yes', 'StreamingTV_No', 'StreamingTV_No internet service', 'StreamingTV_Yes', 'StreamingMovies_No', 'StreamingMovies_No internet service', 'StreamingMovies_Yes', 'Contract_Month-to-month', 'Contract_One year', 'Contract_Two year', 'PaperlessBilling_No', 'PaperlessBilling_Yes', 'PaymentMethod_Bank transfer (automatic)', 'PaymentMethod_Credit 

# ตรวจสอบค่าผิดปกติ เช่น ค่าว่าง ในคอลัมน์ที่มี data type เป็น Object และแทนที่ด้วย 0 ไม่อย่างนั้นจะไม่สามารถทำการ scale or calculate ได้

In [27]:
# เลือก column type object
object_cols = df_dummy.select_dtypes(include='object').columns
print("Object columns:", object_cols)


Object columns: Index(['TotalCharges', 'Churn'], dtype='object')


In [28]:
for col in object_cols:
    # แปลงเป็น numeric, ค่าแปลงไม่ได้จะเป็น NaN
    df_dummy[col] = pd.to_numeric(df_dummy[col], errors='coerce')


In [29]:
# เติม 0 หรือค่าเฉลี่ย
df_dummy.fillna(0, inplace=True)
# หรือถ้าชอบเติม mean
# df_dummy = df_dummy.apply(lambda x: x.fillna(x.mean()))


In [30]:
from scipy.cluster.hierarchy import linkage, dendrogram, fcluster
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

In [31]:
df_dummy['Churn']

2134    0.0
2693    0.0
1679    0.0
5998    0.0
6798    0.0
       ... 
7021    0.0
7026    0.0
7032    0.0
7034    0.0
7041    0.0
Name: Churn, Length: 3738, dtype: float64

Standard

In [32]:
# Scaling
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df_dummy)


In [33]:
X_scaled

array([[-0.48465042, -1.11480881, -1.64605796, ..., -0.82416338,
         1.94266476,  0.        ],
       [-0.48465042,  1.82783407, -1.67541662, ..., -0.82416338,
        -0.51475685,  0.        ],
       [ 2.06334291, -0.86613476, -0.76184407, ..., -0.82416338,
         1.94266476,  0.        ],
       ...,
       [ 2.06334291, -1.11480881,  0.27434407, ...,  1.21335165,
        -0.51475685,  0.        ],
       [-0.48465042,  1.6206057 ,  1.21382132, ..., -0.82416338,
        -0.51475685,  0.        ],
       [ 2.06334291, -0.99047178,  0.2277156 , ..., -0.82416338,
         1.94266476,  0.        ]])

# K - mean Clustering

In [48]:
from sklearn.cluster import KMeans


In [49]:
k = 2  # ตัวอย่างจำนวน cluster
kmeans = KMeans(n_clusters=k, random_state=42)
clusters = kmeans.fit_predict(X_scaled)

In [50]:
df_features['cluster'] = kmeans.labels_

In [51]:
# เปรียบเทียบกับ class จริง
Compare = pd.crosstab(clusters, X_res['Churn'])
Compare

Churn    No   Yes
row_0            
0      1343  1756
1       526   113

In [52]:
X_res['Cluster'] = clusters

Estimate the proportion of High-Risk Avtive

In [65]:

high_risk_act = len(X_res[(X_res['Cluster'] == 0) & (X_res['Churn'] == 'No')])
total = len(X_res)
print('total : ', total )
print(f"High-Risk Active (Target retention): {high_risk_act} คน")
#คำนวณเปอร์เซ็นต์ (Perform Percentage of target retention)
percentage = (high_risk_act / total) * 100
print(f"The proportion of target : {percentage:.2f}", "%")



total :  3738
High-Risk Active (Target retention): 1343 คน
The proportion of target : 35.93 %


# **#Data Insights & Conclusion**

As the figure show us Cluster 0 has 1756 customers that they churned so we need to do something restrain our customers in Cluster 0  and actual class is still No before they decide to say good bye us.

### In other word, 1756 customers already churned (Yes) This means Cluster 0 contain high risk group,where a large portion of customers may churn in the future.

All customers : 3738

The proportion of target to retention about 1343 (35.93%)

การเปรียบเทียบคลาสจริง กับ คลาสที่ได้จากการจัดกลุ่ม ทำให้สามาถประเมินได้ว่าลูกค้าคนใด มีโอกาสที่จะยกเลิกบริการกับเราในอนาคต โดย ลูกค้านั้นจะถูกจัดอยู่ใน cluster 0 แม้จะยังมีคลาสปัจจุบันเป็น NO ก็ตาม แต่มีความเสี่ยงสูงที่จะยกเลิกจากการประเมินค่าใน features ต่างๆ ร่วมกัน

ถ้าเราไปโฟกัสคนที่ Churn ไปแล้ว (Actual Yes) การดึงเขากลับมามีต้นทุนสูงมาก แต่กลุ่ม False Positive คือคนที่ 'พฤติกรรมเริ่มเปลี่ยนแต่ตัวยังอยู่' นี่คือกลุ่มที่เราใช้เงินน้อยที่สุดแต่ได้ผลมากที่สุด



---



In [ ]:
df_features

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
2134  Female              0      No         No       1          Yes   
2693    Male              0      No         No      72          Yes   
1679    Male              1      No        Yes       7          Yes   
5998    Male              0     Yes         No      48          Yes   
6798  Female              0      No         No      13          Yes   
...      ...            ...     ...        ...     ...          ...   
7021    Male              0      No         No      12          Yes   
7026  Female              0      No         No       9          Yes   
7032    Male              1      No         No       1          Yes   
7034  Female              0      No         No      67          Yes   
7041    Male              1     Yes         No       4          Yes   

     MultipleLines InternetService       OnlineSecurity         OnlineBackup  \
2134            No              No  No internet service  No internet service   
2693            No              No  No internet service  No internet service   
1679            No             DSL                   No                   No   
5998            No              No  No internet service  No internet service   
6798            No     Fiber optic                   No                   No   
...            ...             ...                  ...                  ...   
7021            No             DSL                   No                   No   
7026            No             DSL                   No                   No   
7032           Yes     Fiber optic                   No                   No   
7034           Yes     Fiber optic                  Yes                  Yes   
7041           Yes     Fiber optic                   No                   No   

         DeviceProtection          TechSupport          StreamingTV  \
2134  No internet service  No internet service  No internet service   
2693  No internet service  No internet service  No internet service   
1679                   No                   No                   No   
5998  No internet service  No internet service  No internet service   
6798                   No                   No                   No   
...                   ...                  ...                  ...   
7021                   No                  Yes                  Yes   
7026                   No                   No                   No   
7032                   No                   No                   No   
7034                  Yes                   No                  Yes   
7041                   No                   No                   No   

          StreamingMovies        Contract PaperlessBilling  \
2134  No internet service  Month-to-month               No   
2693  No internet service        Two year               No   
1679                   No  Month-to-month              Yes   
5998  No internet service        Two year              Yes   
6798                   No  Month-to-month              Yes   
...                   ...             ...              ...   
7021                   No        One year              Yes   
7026                   No  Month-to-month              Yes   
7032                   No  Month-to-month              Yes   
7034                   No  Month-to-month              Yes   
7041                   No  Month-to-month              Yes   

                  PaymentMethod  MonthlyCharges TotalCharges  cluster  
2134               Mailed check           20.15        20.15        1  
2693  Bank transfer (automatic)           19.30       1414.8        1  
1679               Mailed check           45.75        344.2        0  
5998               Mailed check           20.05         1036        1  
6798               Mailed check           70.15        886.7        0  
...                         ...             ...          ...      ...  
7021           Electronic check           59.80        727.8        0  
7026  Bank transfe

In [ ]:
df_features.shape

(3738, 20)

In [ ]:
accuracy = Compare.apply(max, axis=1).sum() / Compare.to_numpy().sum()
print(accuracy*100)

61.04868913857678


In [ ]:
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency

In [ ]:
def cramers_v(x, y):
    confusion_matrix = pd.crosstab(x, y)
    chi2 = chi2_contingency(confusion_matrix)[0]
    n = confusion_matrix.sum().sum()
    phi2 = chi2/n
    r, k = confusion_matrix.shape
    phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))
    rcorr = r - ((r-1)**2)/(n-1)
    kcorr = k - ((k-1)**2)/(n-1)
    return np.sqrt(phi2corr / min((kcorr-1), (rcorr-1)))

categorical_cols = df_features.columns

for col in categorical_cols:
    print(f"{col}: {cramers_v(df_features['cluster'], df_features[col]):.3f}")




gender: 0.000
SeniorCitizen: 0.170
Partner: 0.015
Dependents: 0.149
tenure: 0.046
PhoneService: 0.146
MultipleLines: 0.293
InternetService: 1.000
OnlineSecurity: 1.000
OnlineBackup: 1.000
DeviceProtection: 1.000
TechSupport: 1.000
StreamingTV: 1.000
StreamingMovies: 1.000
Contract: 0.264
PaperlessBilling: 0.317
PaymentMethod: 0.370
MonthlyCharges: 0.774
TotalCharges: 0.164
cluster: 0.999


## Cramér’s V หากมีค่าสูงๆ แปลว่า Feature นั้นมีอิทธิพลต่อการจัดกลุ่มมาก

Cluster 0: class Yes (ออก) คนส่วนใหญ่เป็น Yes (1756) แต่เมื่อเปรียบเทียบกับคลาสจริง ยังคงมี class No รวมอยู่มาก

Cluster 1: class No (ไม่ออก) คนส่วนใหญ่เป็น No (526) และเมื่อเปรียบเทียบกับคลาสจริง Yes มีน้อย (113)

## **Check Columns (No need to)**

In [ ]:
df_dummy.nunique()

SeniorCitizen                                 2
tenure                                       73
MonthlyCharges                             1297
TotalCharges                               3500
gender_Female                                 2
gender_Male                                   2
Partner_No                                    2
Partner_Yes                                   2
Dependents_No                                 2
Dependents_Yes                                2
PhoneService_No                               2
PhoneService_Yes                              2
MultipleLines_No                              2
MultipleLines_No phone service                2
MultipleLines_Yes                             2
InternetService_DSL                           2
InternetService_Fiber optic                   2
InternetService_No                            2
OnlineSecurity_No                             2
OnlineSecurity_No internet service            2
OnlineSecurity_Yes                            2
OnlineBackup_No                               2
OnlineBackup_No internet service              2
OnlineBackup_Yes                              2
DeviceProtection_No                           2
DeviceProtection_No internet service          2
DeviceProtection_Yes                          2
TechSupport_No                                2
TechSupport_No internet service               2
TechSupport_Yes                               2
StreamingTV_No                                2
StreamingTV_No internet service               2
StreamingTV_Yes                               2
StreamingMovies_No                            2
StreamingMovies_No internet service           2
StreamingMovies_Yes                           2
Contract_Month-to-month                       2
Contract_One year                             2
Contract_Two year                             2
PaperlessBilling_No                           2
PaperlessBilling_Yes                          2
PaymentMethod_Bank transfer (automatic)       2
PaymentMethod_Credit card (automatic)         2
PaymentMethod_Electronic check                2
PaymentMethod_Mailed check                    2
Churn                                         2
dtype: int64

In [ ]:
df_dummy.columns

Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges',
       'gender_Female', 'gender_Male', 'Partner_No', 'Partner_Yes',
       'Dependents_No', 'Dependents_Yes', 'PhoneService_No',
       'PhoneService_Yes', 'MultipleLines_No',
       'MultipleLines_No phone service', 'MultipleLines_Yes',
       'InternetService_DSL', 'InternetService_Fiber optic',
       'InternetService_No', 'OnlineSecurity_No',
       'OnlineSecurity_No internet service', 'OnlineSecurity_Yes',
       'OnlineBackup_No', 'OnlineBackup_No internet service',
       'OnlineBackup_Yes', 'DeviceProtection_No',
       'DeviceProtection_No internet service', 'DeviceProtection_Yes',
       'TechSupport_No', 'TechSupport_No internet service', 'TechSupport_Yes',
       'StreamingTV_No', 'StreamingTV_No internet service', 'StreamingTV_Yes',
       'StreamingMovies_No', 'StreamingMovies_No internet service',
       'StreamingMovies_Yes', 'Contract_Month-to-month', 'Contract_One year',
       'Contract_Two year', '

In [ ]:
df_dummy.dtypes

SeniorCitizen                                int64
tenure                                       int64
MonthlyCharges                             float64
TotalCharges                               float64
gender_Female                                 bool
gender_Male                                   bool
Partner_No                                    bool
Partner_Yes                                   bool
Dependents_No                                 bool
Dependents_Yes                                bool
PhoneService_No                               bool
PhoneService_Yes                              bool
MultipleLines_No                              bool
MultipleLines_No phone service                bool
MultipleLines_Yes                             bool
InternetService_DSL                           bool
InternetService_Fiber optic                   bool
InternetService_No                            bool
OnlineSecurity_No                             bool
OnlineSecurity_No internet service            bool
OnlineSecurity_Yes                            bool
OnlineBackup_No                               bool
OnlineBackup_No internet service              bool
OnlineBackup_Yes                              bool
DeviceProtection_No                           bool
DeviceProtection_No internet service          bool
DeviceProtection_Yes                          bool
TechSupport_No                                bool
TechSupport_No internet service               bool
TechSupport_Yes                               bool
StreamingTV_No                                bool
StreamingTV_No internet service               bool
StreamingTV_Yes                               bool
StreamingMovies_No                            bool
StreamingMovies_No internet service           bool
StreamingMovies_Yes                           bool
Contract_Month-to-month                       bool
Contract_One year                             bool
Contract_Two year                             bool
PaperlessBilling_No                           bool
PaperlessBilling_Yes                          bool
PaymentMethod_Bank transfer (automatic)       bool
PaymentMethod_Credit card (automatic)         bool
PaymentMethod_Electronic check                bool
PaymentMethod_Mailed check                    bool
Churn                                      float64
dtype: object